### PR TITLE
APPS-1378 dxcint async tests

### PR DESCRIPTION
**Pull request by JanProvaznik**
In the end there is no asynchronous code, only concurrent. Since the tests are independent, I simply gave a thread to each one (preemptive multithreading) and they can use DXJob.wait_for_done without interfering with each other. Performance should be OK unless there are >1000 tests.

### CODE REVIEW

File: dxcint/dxcint/TestDiscovery.py [hunk 1]
```diff
@@ -19,6 +19,7 @@ class TestDiscoveryError(Exception):
 
 
 class TestDiscovery(object):
+    __test__ = False
```
Comment (Gvaihir) [inline dxcint/dxcint/TestDiscovery.py:22]: good catch

File: dxcint/dxcint/RegisteredTest.py [hunk 2]
```diff
@@ -132,8 +134,8 @@ def _run_executable(self) -> str:
 
         Returns: str. Execution ID (analysis or job)
         """
-        exec_desc = self._run_executable_inner().describe()
-        return exec_desc.get("id")
+        execution_desc = self._run_executable_inner().describe()
+        return execution_desc.get("id")
```
Comment (Gvaihir) [inline dxcint/dxcint/RegisteredTest.py:144]: ok, more readable

File: dxcint/dxcint/RegisteredTest.py [hunk 3]
```diff
@@ -32,8 +32,8 @@ def __init__(
         self._test_name = test_name
         self._language = rm_prefix(os.path.basename(src_file), test_name).strip(".")
         self._messenger = None
```
Comment (Gvaihir) [inline dxcint/dxcint/RegisteredTest.py:28]: self._create_messenger()

File: dxcint/dxcint/cli.py [hunk 4]
```diff
@@ -69,8 +71,21 @@ def integration(
         dependencies=test_discovery.discover_dependencies()
     )
     _ = terraform.build()
-    for test in registered_tests:
-        passed = test.test_result
+
+    # simply give every test a thread, no need for cooperative multitasking,
+    # tests are independent, they will print correctly progress to logger
```
Comment (Gvaihir) [inline dxcint/dxcint/cli.py]:
```suggestion
    # tests are independent, they will correctly print the progress to the logger
```

File: dxcint/dxcint/cli.py [hunk 5]
```diff
@@ -108,7 +123,8 @@ def add(
         CATEGORY: Test category name. Usually a team-defined category that reflects a type of the test.
                 Check dxcint/config/{SUITE_FILE}.json, where the keys are category names.
     """
-    test_context = Context(project="dxCompiler_playground", repo_root=dxc_repository_root)
+    test_context = Context(project="dxCompiler_playground",
+                           repo_root=dxc_repository_root)
```
Comment (Gvaihir) [inline dxcint/dxcint/cli.py]: Pls follow [black](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#parentheses). I'll create a ticket for you to add pre-commit action for black and flake8 for the python code.

File: dxcint/dxcint/cli.py [hunk 6]
```diff
@@ -69,8 +71,21 @@ def integration(
...
+
+    # simply give every test a thread, no need for cooperative multitasking,
+    # tests are independent, they will print correctly progress to logger
+    with ThreadPoolExecutor(max_workers=len(registered_tests)+5) as executor:
+        future_to_execute_tests = \
+            {executor.submit(registered_test.get_test_result) 
+             for registered_test in registered_tests}
+        results: List[bool] = [f.result()
+                   for f in futures.as_completed(future_to_execute_tests)]
+    failures = results.count(False)
```
Comment (Gvaihir) [inline dxcint/dxcint/cli.py]: `num_failures` or `failure_counter`

File: dxcint/tests/conftest.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 from dxcint.Context import Context
 from dxcint.Dependency import DependencyFactory
 from dxcint.Terraform import Terraform
+from .test_RegisteredTest import build_executable_wdl, registered_test_wdl
```
Comment (Gvaihir) [inline dxcint/tests/conftest.py]: please no relative imports. Also conftest should not import anything to potentially avoid cyclic imports. Just move those fixtures to conftest

File: dxcint/dxcint/cli.py [hunk 8]
```diff
@@ -69,8 +71,21 @@ def integration(
...
-        passed = test.test_result
+
+    # simply give every test a thread, no need for cooperative multitasking,
+    # tests are independent, they will print correctly progress to logger
+    with ThreadPoolExecutor(max_workers=len(registered_tests)+5) as executor:
+        future_to_execute_tests = \
+            {executor.submit(registered_test.get_test_result) 
+             for registered_test in registered_tests}
+        results: List[bool] = [f.result()
+                   for f in futures.as_completed(future_to_execute_tests)]
```
Comment (Gvaihir) [inline dxcint/dxcint/cli.py]: pls follow black

File: dxcint/dxcint/testclasses/ExpectedFailure.py [hunk 9]
```diff
@@ -12,13 +11,9 @@ def __init__(self, src_file: str, category: str, test_name: str, context: Contex
...
     def _run_executable(self) -> str:
         execution = self._run_executable_inner()
-        try:
-            execution.wait_on_done()
-        except dxpy.DXJobFailureError:
-            pass
         return execution.describe().get("id")
 
-    def _validate(self) -> Dict:
+    async def _validate(self) -> Dict:
```
Comment (Gvaihir) [inline dxcint/dxcint/testclasses/ExpectedFailure.py]: why making it a coroutine? You are not using `await`. Or is it a carry-over form other implementation attempts?

File: dxcint/dxcint/Messenger.py [hunk 10]
```diff
@@ -1,60 +1,70 @@
...
-        """
-        Class to maintain the status of current test
-        """
-        self._state = self.NOT_STARTED
+    def __init__(self, test_name: str, job_id: str,
+                 variant: Optional[object] = None, interval: Optional[int] = 20):
+        self._test_name = test_name
+        self._variant = variant if variant else ""
+        self._state: Union[State, str] = State.NOT_DONE
+        self._job: Union[dxpy.DXAnalysis, dxpy.DXJob] = None
```
Comment (Gvaihir) [inline dxcint/dxcint/Messenger.py]: Redundant because next line

File: dxcint/dxcint/Messenger.py [hunk 11]
```diff
@@ -1,60 +1,70 @@
...
+        Most common usage is in _validate method of RegisteredTest descendant.
 
-    def wait_for_completion(self) -> str:
-        logging.info(f"Waiting for completion of the test `{self._test_name}` with input variant {self._variant}")
-        test_name_with_variant = ".".join([self._test_name, self._variant]).strip(".")
+        returns: State.FINISHED if the job/analysis ran to completion, State.FAIL if it failed or was terminated.
+        """
+        logging.info(
+            f"Waiting for completion of the test `{self._test_name}` with input variant {self._variant}")
+        test_name_with_variant = ".".join(
```
Comment (Gvaihir) [inline dxcint/dxcint/Messenger.py]: f-string better?

File: dxcint/dxcint/Messenger.py [hunk 12]
```diff
@@ -1,60 +1,70 @@
...
+    def __init__(self, test_name: str, job_id: str,
+                 variant: Optional[object] = None, interval: Optional[int] = 20):
+        self._test_name = test_name
+        self._variant = variant if variant else ""
+        self._state: Union[State, str] = State.NOT_DONE
+        self._job: Union[dxpy.DXAnalysis, dxpy.DXJob] = None
+        if 'analysis-' in job_id:
+            self._job = dxpy.DXAnalysis(job_id)
+        else:
+            self._job = dxpy.DXJob(job_id)
```
Comment (Gvaihir) [inline dxcint/dxcint/Messenger.py]: check `dxpy.bindings.dxdataobject_functions.get_handler` - maybe no need for a conditional

File: dxcint/dxcint/Messenger.py [hunk 13]
```diff
@@ -1,60 +1,70 @@
...
+            self._state = State.FINISHED
         except dxpy.DXJobFailureError:
-            logging.info(f"Analysis {test_name_with_variant} failed. This failure will verified")
-            self._status.state = Status.FAIL
+            logging.info(
+                f"Analysis {test_name_with_variant} failed. This failure will verified")
+            self._state = State.FAIL
+        return self.state
+
+    def query_state(self, extended: bool = False) -> Union[State, str]:
```
Comment (Gvaihir) [inline dxcint/dxcint/Messenger.py:63]: good, handy

File: dxcint/dxcint/Messenger.py [hunk 14]
```diff
@@ -1,60 +1,70 @@
...
+        """Queries the state of the job and returns it in the form of a State enum."""
```
Comment (Gvaihir) [inline dxcint/dxcint/Messenger.py]: If method has args and returns stuff please include those in the docstring

File: dxcint/tests/test_Messenger.py [hunk 15]
```diff
@@ -1,9 +1,45 @@
...
+        test_futures = {
+            executor.submit(test1.get_test_result),
+            executor.submit(test2.get_test_result)
+        }
+        results = [f.result() for f in as_completed(test_futures)]
+        history1 = h1.result()
+        history2 = h2.result()
+
+    assert results == [False, False]
+    assert State.NOT_DONE in history1
```
Comment (Gvaihir) [inline dxcint/tests/test_Messenger.py]: Can you test for the order at which those statuses appear to confirm concurrency? For example - create a dict with both histories and update it with the status changes. Then check that test 1 did not wait for test 2 to finish

**Review by Gvaihir — CHANGES_REQUESTED** [review]
Great job! Requested changes. Will add a ticket for code linter for this package and precommits.
EDIT: APPS-1397

File: dxcint/dxcint/RegisteredTest.py (hunk 3 above)
Comment (JanProvaznik) [reply dxcint/dxcint/RegisteredTest.py:28]: No, the messenger is and should be created only after the test is ran, which is not the same as when the class is created.

File: dxcint/dxcint/testclasses/ExpectedFailure.py (hunk 9 above)
Comment (JanProvaznik) [reply dxcint/dxcint/testclasses/ExpectedFailure.py]: thanks for spotting it, forgot it there, deleted in the next commit

File: dxcint/tests/test_Messenger.py (hunk 15 above)
Comment (JanProvaznik) [reply dxcint/tests/test_Messenger.py]: Added a simple check for that the jobs share some time in NOT_DONE state but beyond that the workers are quite unpredictable, so one waiting and one being done is not impossible even if they are submitted at the same time.

File: dxcint/dxcint/RegisteredTest.py (hunk 3 above)
Comment (Gvaihir) [reply dxcint/dxcint/RegisteredTest.py:28]: right, there's a dependency on job ID

**Review by Gvaihir — APPROVED** [review]
Looks good!

**Pull request merged**